src/mc_sketch_alg.cpp

```diff
@@ -829,42 +829,92 @@ SpanningForest MCSketchAlg::calc_spanning_forest(size_t graph_id) {
   return ret;
 }
 
+void MCSketchAlg::filter_sf_edges(SpanningForest &sf, size_t graph_id) {
+  auto start = std::chrono::steady_clock::now();
+
+  dsu_valid = false;
+  shared_dsu_valid = false;
+
+  const std::vector<Edge> &edges = sf.get_sorted_adjacency();
+
+#pragma omp parallel
+  {
+    size_t thr_id = omp_get_thread_num();
+    size_t num_threads = omp_get_num_threads();
+
+    std::pair<size_t, size_t> partition = get_ith_partition(edges.size(), thr_id, num_threads);
+    size_t start = partition.first;
+    size_t end = partition.second;
+
+    // check if we collide with previous thread. If so lock and apply those updates.
+    if (start > 0 && edges[start].src == edges[start - 1].src) {
+      sketches[(graph_id * num_vertices) + edges[start].src]->mutex.lock();
+      size_t orig_start = start;
+      while (edges[start].src == edges[orig_start].src) {
+        Edge edge = edges[start];
+        // TODO: Need to address that this should only update a particular subgraph!
+        sketches[(graph_id * num_vertices) + edge.src]->update(static_cast<vec_t>(concat_pairing_fn(edge.src, edge.dst)));
+        ++start;
+      }
+
+      sketches[(graph_id * num_vertices) + edges[orig_start].src]->mutex.unlock();
+    }
+
+    // check if we collide with next thread. If so lock and apply those updates.
+    if (end < edges.size() && edges[end - 1].src == edges[end].src) {
+      sketches[(graph_id * num_vertices) + edges[end - 1].src]->mutex.lock();
+      size_t orig_end = end;
+      while (edges[end - 1].src == edges[orig_end - 1].src) {
+        Edge edge = edges[end - 1];
+        sketches[(graph_id * num_vertices) + edge.src]->update(static_cast<vec_t>(concat_pairing_fn(edge.src, edge.dst)));
+        --end;
+      }
+
+      sketches[(graph_id * num_vertices) + edges[orig_end].src]->mutex.unlock();
+    }
+    
+    for (size_t i = start; i < end; i++) {
+      Edge edge = edges[i];
+      sketches[(graph_id * num_vertices) + edge.src]->update(static_cast<vec_t>(concat_pairing_fn(edge.src, edge.dst)));
+    }
+  }
+
+  delete_time += std::chrono::steady_clock::now() - start;
+}
+
 std::vector<SpanningForest> MCSketchAlg::calc_disjoint_spanning_forests(size_t graph_id, size_t k) {
+  sf_query_start = std::chrono::steady_clock::now();
   std::vector<SpanningForest> SFs;
+  std::chrono::steady_clock::time_point start;
   size_t max_rounds = 0;
 
   for (size_t i = 0; i < k; i++) {
-    SpanningForest sf = calc_spanning_forest(graph_id);
-
+    start = std::chrono::steady_clock::now();
+    SFs.push_back(calc_spanning_forest());
```

Comment:
Note to self: I think this is where the bug is. Should include graph_id here! Check if this issue crops up elsewhere.